xonsh/imphooks.py

```diff
@@ -0,0 +1,78 @@
+"""Import hooks for importing xonsh source files. This module registers
+the hooks it defines when it is imported.
+"""
+from __future__ import print_function, unicode_literals
+import os
+import sys
+import builtins
+from importlib.machinery import ModuleSpec
+from importlib.abc import MetaPathFinder, SourceLoader
+
+from xonsh.tools import string_types
+from xonsh.execer import Execer
+
+class XonshImportHook(MetaPathFinder, SourceLoader):
+    """Implements the import hook for xonsh source files."""
+
+    def __init__(self, *args, **kwargs):
+        super(XonshImportHook, self).__init__(*args, **kwargs)
+        self._filenames = {}
+        self._execer = None
+
+    @property
+    def execer(self):
+        if hasattr(builtins, '__xonsh_execer__'):
+            execer = builtins.__xonsh_execer__
+            if self._execer is not None:
+                self._execer = None
+        elif self._execer is None:
+            self._execer = execer = Execer(unload=False)
+        else:
+            execer = self._execer
+        return execer
+    #
+    # MetaPathFinder methods
+    #
+    def find_spec(self, fullname, path, target=None):
+        """Finds the spec for a xonsh module if it exists."""
+        spec = None
+        path = sys.path if path is None else path
```

Comment:
Also, this seems not _quite_ to work after a little more testing.  With a directory structure:

```
* testdir
|---* __init__.py
|---* test_script.py
|---* test_script2.py
```

It took two tries for the import to go through.

```
>>> import testdir
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/site-packages/xonsh/shell.py", line 123, in default
    self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
  File "/usr/local/lib/python3.4/site-packages/xonsh/execer.py", line 115, in exec
    return exec(code, glbs, locs)
  File "<xonsh-code>", line 1, in <module>
ImportError: No module named 'testdir'
>>> import testdir
>>> testdir
<module 'testdir' from '/home/testuser/testdir/__init__.py'>
```

Putting `testdir` in my path worked fine.